hive-hooks/src/main/java/com/airbnb/reair/hive/hooks/MetastoreAuditLogHook.java

```diff
@@ -0,0 +1,382 @@
+   */
+  @Override
+  public void onAlterPartition(AlterPartitionEvent event) throws MetaException {
+    try {
+
+      // Table is invariant and thus arbitrary choice between old and new.
+      Table table = new Table(
+        event.getOldPartition().getDbName(),
+        event.getOldPartition().getTableName()
+      );
+
+      table.setPartCols(event.getOldPartition().getSd().getCols());
+
+      Set<ReadEntity> readEntities = new HashSet<>();
+      readEntities.add(new ReadEntity(table));
+
+      readEntities.add(
+        new ReadEntity(new Partition(table, event.getOldPartition()))
+      );
+
+      Set<WriteEntity> writeEntities = new HashSet<>();
+      writeEntities.add(new WriteEntity(table, WriteType.INSERT));
+
+      writeEntities.add(
+        new WriteEntity(
+          new Partition(table, event.getNewPartition()),
+          WriteType.INSERT
+        )
+      );
+
+      run(readEntities, writeEntities, HiveOperation.ALTERPARITTION);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Listener which fires when a database (schema) is created.
+   *
+   * The event only provides the database which was created, hence for auditing
+   * purposes the read/write differential is the non-presence and presence of
+   * the database respectively.
+   *
+   * @param event The create database event
+   */
+  @Override
+  public void onCreateDatabase(CreateDatabaseEvent event) throws MetaException {
+    try {
+      Set<ReadEntity> readEntities = new HashSet<>();
+      Set<WriteEntity> writeEntities = new HashSet<>();
+      writeEntities.add(new WriteEntity(event.getDatabase(), WriteType.INSERT));
+
+      run(readEntities, writeEntities, HiveOperation.CREATEDATABASE);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Listener which fires when a database (schema) is dropped.
+   *
+   * The event only provides the database which was dropped, hence for auditing
+   * purposes the read/write differential is the presence and non-presence of
+   * the database respectively.
+   *
+   * @param event The drop database event
+   */
+  @Override
+  public void onDropDatabase(DropDatabaseEvent event) throws MetaException {
+    try {
+      Set<ReadEntity> readEntities = new HashSet<>();
+      readEntities.add(new ReadEntity(event.getDatabase()));
+      Set<WriteEntity> writeEntities = new HashSet<>();
+
+      run(readEntities, writeEntities, HiveOperation.DROPDATABASE);
+    } catch (Exception e) {
+      throw new MetaException(e.getMessage());
+    }
+  }
+
+  /**
+   * Runs the individual metastore audit log modules that make up this hook.
+   *
+   * @param readEntities The entities that were read by the query
+   * @param writeEntities The entities that were written by the query
+   * @param hiveOperation The Hive operation
+   * @throws Exception if there's an error running the modules
+   */
+  private void run(
```

Comment:
Can we be more specific than throwing Exception? The signature of the original hook was due to the API.